---
output:
    github_document:
        html_preview: false
---

<!-- README.md is generated from README.Rmd. Please edit that file -->

```{r, include = FALSE}
knitr::opts_chunk$set(
  collapse = TRUE,
  comment = "#>",
  fig.path = "man/figures/README-",
  out.width = "100%",
  eval = FALSE
)
```

# CAMDAC

Copy-number Aware Methylation Deconvolution Analysis of Cancer (CAMDAC) is a tool that deconvolves pure tumor methylation from bulk tumor sequencing data ([Larose Cadieux et al., 2022, bioXriv](https://www.biorxiv.org/content/10.1101/2020.11.03.366252v2)).

This branch describes CAMDAC for whole genome bisulfite sequencing (WGBS) data. To run CAMDAC on Reduced Representation Bisulfite Sequencing (RRBS) data, visit the [VanLoo-lab/CAMDAC main branch](https://github.com/VanLoo-lab/CAMDAC/tree/main).

<!-- badges: start -->
<!-- badges: end -->

## Installation

## Documentation

View the full documentation at [https://vanloo-lab.github.io/CAMDAC/](https://vanloo-lab.github.io/CAMDAC/).

## Quickstart

CAMDAC can be installed from an R console:

``` r
remotes::install_github("VanLoo-lab/CAMDAC@wgbs")
CAMDAC::download_pipeline_files("wgbs")
```

To run the tumor-normal deconvolution pipeline with test data:

```{r example}
library(CAMDAC)
tumor = CamConfig("P1_T1", bam_file = "tumor.bam")
normal = CamConfig("P1_N1", bam_file = "normal.bam")
config = CamConfig(outdir="./results", bsseq="wgbs", bsseq_lib="pe", build="hg38")
CamPipeline(tumor, normal, config)
```
