---
title: "allele-specific"
output: rmarkdown::html_vignette
vignette: >
  %\VignetteIndexEntry{allele-specific}
  %\VignetteEngine{knitr::rmarkdown}
  %\VignetteEncoding{UTF-8}
---

```{r, include = FALSE}
knitr::opts_chunk$set(
  collapse = TRUE,
  comment = "#>"
)
```

## Â Allele-specific methylation (ASM) analysis

CAMDAC can be used to detect allele-specific methylation (ASM) by phasing CpGs to heterozygous SNPs and deconvolving bulk methylation rates per allele. 

This tutorial steps through the ASM analysis pipeline. Briefly:
1. Count CpG methylation on tumor and normal at sites phased to SNP loci.
2. Deconvolve methylation on tumor **per haplotype** using the normal
3. Assign allele-specific copy number state **per CpG** using the bulk tumor solution
4. Call allele-specific differential methylation within samples
5. Call allele-specific differential methylation between samples

Results from this pipeline are found in the results directory under 'PATIENT/AlleleSpecific'. The following files are generated:
- *asm_counts.csv.gz - The number of reads supporting each allele at each CpG
- *asm_hap_stats.csv.gz - Summary statistics for each phased SNP
- *asm_phase_map.csv.gz - A mapping if read-CpG-SNP phasing
- *snps.txt - The heterozygous SNP loci input for ASM analysis 
- *cna.csv - The tumor allele-specific copy number solution. See format in `vignettes("pipeline")`.
- *asm_cg_cn.csv - Phased CpG copy number state
- *asm_m.csv - Phased CpG bulk methylatio
- *asm_m_pure.csv - Phased CpG pure methylation
- *asm_m_within_diff.csv - Within-sample differential methylation
- *asm_m_between_diff.csv - Between-sample differential methylation
- *asm_meta.csv - Description of all input parameters for CAMDAC ASM run.

## CAMDAC-ASM from BAM files

The `asm_pipeline()` function runs CAMDAC-ASM analysis by generates the allele-specific copy number solution and heterozygous SNP loci, followesd by deconvolution and differential ASM analysis:

```{r}
b_tumor <- system.file("testdata", "tumor.bam", package = "CAMDAC")
b_normal <- system.file("testdata", "normal.bam", package = "CAMDAC")

tumor <- CamSample(id="T", sex="XY", bam=b_tumor)
normal <- CamSample(id="N", sex="XY", bam=b_normal)
config <- CamConfig(outdir="./results", bsseq="wgbs", lib="pe", build="hg38", n_cores=10)

asm_pipeline(
    tumor=tumor,
    germline=normal,
    infiltrates=normal,
    origin=normal,
    config=config
)
```

The ASM results are placed in a PATIENT/AlleleSpecific directory with the following files:
- 

## TODO: CAMDAC-ASM from external inputs

To run the ASM pipeline without BAM files, CAMDAC requires:
- Each CamSample object has SNP loci
- The tumor CamSample object has an allele-specific CNA solution
- All CamSample objects have BAM files available for phasing

CAMDAC-ASM requires a file of heterozygous SNP loci against which CpGs will be phased. This is a tab-delimited file with a header containing four fields:

| Field | Description |
| ---- | ------------ |
| chrom | Chromosome name |
| pos | SNP loci position |
| ref | The reference allele (A/C/T/G) |
| alt | The alternate SNP allele (A/C/T/G) | 

## ASM analysis after CAMDAC main pipeline 

TODO

## ASM analysis with external inputs

First, attach your SNP loci file to the tumor object with `attach_output()`, then run `asm_pipeline()`:
```{r}
# Setup CAMDAC samples
tumor <- CamSample(id="tumor", sex="XY", bam=b_tumor)
normal <- CamSample(id="normal", sex="XY", bam=b_normal)
config <- CamConfig(outdir="./results", ref="./pipeline_files", bsseq="wgbs", lib="pe", cores=10)

# Add SNPs
asm_snps_file <- system.file("testdata", "test_het_snps.tsv", package = "CAMDAC")
attach_output(tumor, config, "asm_snps", asm_snps_file)
attach_output(normal, config, "asm_snps", asm_snps_file)
```

Next, CAMDAC requires the allele-specific copy number solution from the tumor, attached as follows:
```{r}
cna_file <- system.file("testdata", "test_cna.tsv", package = "CAMDAC")
attach_output(tumor, config, "cna", cna_file)
```

Finally, run the allele-specific methylation pipeline:
```{r}
asm_pipeline(
  tumor=tumor,
  infiltrates=normal,
  origin=normal,
  config=config
)
```

## Using SNPs called from previous CAMDAC runs

If you have already run the CAMDAC pipeline in tumor-normal mode, then the tumor object's SNP files will be used by default.
