---
title: "deconvolution"
output: rmarkdown::html_vignette
vignette: >
  %\VignetteIndexEntry{deconvolution}
  %\VignetteEngine{rmarkdown::render}
  %\VignetteEncoding{UTF-8}
---

```{r, include = FALSE}
knitr::opts_chunk$set(
  collapse = TRUE,
  comment = "#>"
)
```

## Introduction

Here, we demonstrate the simplest case for CAMDAC deconvolution using a bulk tumor with adjacent normal tissue from the same patient. We can define CAMDAC inputs using two functions:

- `create_camdac_config()` sets the configuration for camdac pipeline functions.
- `create_camdac_sample()` sets details for each biological sample.

After which, we pass these input objects to `pipline_*` functions. This vignette will walk you through running the `pipeline_tumor_normal()` workflow.

Load the CAMDAC library:

```{r camdac_load}
suppressPackageStartupMessages(
  suppressWarnings(
    library(CAMDAC)
  )
)
```

## Configuration

### Reference files

CAMDAC requires a set of pre-built reference files. These files are stored at a dropbox endpoint as tar.gz files. You can download these files with the command `CAMDAC::download_pipeline_files("wgbs")"`. 

CAMDAC pipeline functions search for the unpacked reference files in the following locations:

1. The directory passed when creating the CAMDAC config object (see camdac_refs argument in `create_camdac_config`)
1. The location defined by the environment variable CAMDAC_PIPELINE_FILES.
1. The current working directory

If the download command fails, you can manually download the reference files from the endpoints listed in `inst/extdata/pipeline_files_urls.txt`.

### Configuration object

Once reference files are downloaded, load the CAMDAC library and set up the configuration:

```{r config_setup}
config = create_camdac_config(
  # camdac_refs = "./camdac_refs", # Optional: Pass the location of pipeline files 
  outdir="./results", # Path for outputs
  bsseq="wgbs", # WGBS
  build="hg38", # Reference
  bsseq_lib="pe", # Paired end
  n_cores = 10 # Cores for parallel processing
)
```

## Deconvolution

Next, define CAMDAC samples from the bulk tumor and normal BAM files. Test BAM files are provided with CAMDAC to verify your installation but please be aware that **these are not whole genome bisulfite sequencing data**, therefore the outputs from these files are not meaningful.

```{r camdac_sample}
# Get test BAM
tumor_bam = system.file("extdata", "NA18939_test_v3.bam", package="CAMDAC")
normal_bam = system.file("extdata", "NA20502_test_v3.bam", package="CAMDAC")

# Set CAMDAC tumor sample
tumour <- create_camdac_sample(
  patient_id="P1",
  patient_sex="XX",
  sample_id="T",
  sample_type="tumour",
  bam_file=tumor_bam
)

# Set CAMDAC normal sample
normal <- create_camdac_sample(
  patient_id="P1",
  patient_sex="XX",
  sample_id="N",
  sample_type="normal",
  bam_file=normal_bam
)
```

Run the CAMDAC tumor-normal pipeline. This analysis takes 1 hour using the test dataset, and approximately 8 hours on a full WGBS dataset.

```{r deconvolve, eval=FALSE}
pipeline_tumor_normal(tumor, normal, config)
```

In this pipeline, the adjacent normal methylation profile is used to select germline heterozygous SNPs for copy number calling, to estimate the infiltrating cell methylation, and as a proxy for the cell of origin in differential methylation calling.

## Results

CAMDAC outputs are written in the directory given by `config$outdir` in the format `PATIENT/DATASET/SAMPLE/`:
```
└── P1
    ├── Allelecounts
    │   ├── N
    │   │   └── P1.N.SNPs.CpGs.all.sorted.csv.gz
    │   └── T
    │       └── P1.T.SNPs.CpGs.all.sorted.csv.gz
    ├── Copynumber
    │   └── T
    │       ├── battenberg # (Subset of files shown)
    │           ├── P1-T_subclones.txt
    │           └── P1-T_BattenbergProfile_average.png
    │       ├── P1.T.ACF.and.ploidy.txt
    │       ├── P1.T.ascat.bc.qs
    │       ├── P1.T.ascat.frag.qs
    │       ├── P1.T.ascat.output.qs
    │       ├── P1.T.ASPCF.png
    │       ├── P1.T.germline.png
    │       ├── P1.T.SNPs.csv.gz
    │       ├── P1.T.ASCATprofile.png
    │       ├── P1.T.BAF.PCFed.txt
    │       ├── P1.T.LogR.PCFed.txt
    │       ├── P1.T.rawprofile.png
    │       ├── P1.T.sunrise.png
    │       └── P1.T.tumour.png
    └── Methylation
        ├── N
        │   └── P1.N.m.csv.gz
        └── T
            ├── P1.T.CAMDAC_annotated_DMRs.fst
            ├── P1.T.CAMDAC_results_per_CpG.fst
            ├── P1.T.m.csv.gz
            └── P1.T.pure.qs

```