---
title: "output"
output: rmarkdown::html_vignette
vignette: >
  %\VignetteIndexEntry{output}
  %\VignetteEngine{rmarkdown::render}
  %\VignetteEncoding{UTF-8}
---

```{r, include = FALSE}
knitr::opts_chunk$set(
  collapse = TRUE,
  comment = "#>",
  eval=FALSE
)
```

CAMDAC outputs are written in the directory given by `config$outdir` in the format `PATIENT/DATASET/SAMPLE/`:

```
└── P
    ├── Allelecounts
    │   ├── N
    │   │   └── P.N.SNPs.CpGs.all.sorted.csv.gz
    │   └── T
    │       └── P.T.SNPs.CpGs.all.sorted.csv.gz
    ├── Copynumber
    │   ├── N
    │   │   └── P.N.SNPs.csv.gz
    │   └── T
    │       ├── ascat
    │       ├── battenberg
    │       ├── P.T.cna.txt
    │       ├── P.T.SNPs.csv.gz
    │       └── P.T.tnSNP.csv.gz
    └── Methylation
        ├── N
        │   └── P.N.m.csv.gz
        └── T
            ├── P.T.CAMDAC_annotated_DMRs.fst
            ├── P.T.CAMDAC_results_per_CpG.fst
            ├── P.T.m.csv.gz
            └── P.T.pure.csv.gz
```

| File | Description |
|------|-------------|
| `P.T.SNPs.CpGs.all.sorted.csv.gz` | Allele counts for a sample. Generated by processing BAM file |
| `P.T.SNPs.csv.gz` | SNP counts for a sample. |
| `P.T.cna.txt` | CAMDAC CNA result |
| `P.T.m.csv.gz` | Bulk methylation data |
| `P.T.m.pure.csv.gz` | CAMDAC-deconvolved methylation data |
| `P1.T.CAMDAC_results_per_CpG.fst` | CAMDAC differentially methylated cytosines |
| `P1.T.CAMDAC_annotated_DMRs.fst` | CAMDAC differentially methylated regions |

It is possible to manually override outputs for runs. See `vignette("questions")` for more details.
