---
title: "pipeline"
output: rmarkdown::html_vignette
vignette: >
  %\VignetteIndexEntry{pipeline}
  %\VignetteEngine{rmarkdown::render}
  %\VignetteEncoding{UTF-8}
---

```{r, include = FALSE}
knitr::opts_chunk$set(
  collapse = TRUE,
  comment = "#>",
  eval = FALSE
)
```

The CAMDAC pipeline accepts raw data for each of the following `CamSample` objects:

1. `tumor` - The bulk tumor sample you  to deconvolve
2. `germline` (optional) - Patient-matched normal data for allele-specific copy number calling. Alternatively, see `using external CNA solutions` below.
3. `infiltrates` - Bulk methylome representing the normal infiltrating cells for deconvolution. E.g. normal adjacent tissue.
4. `origin` (optional) - Bulk methylome for differential methylation calling, representing the normal cell type from which the tumour originated.

The same normal sample may be  given as the germline, infiltrates or origin, depending on your experimental design.

## Run the CAMDAC pipeline

```{r pipeline}
library(CAMDAC)

# Load test data
b_tumor <- system.file("testdata", "tumor.bam", package = "CAMDAC")
b_normal <- system.file("testdata", "normal.bam", package = "CAMDAC")

tumor <- CamSample(id="T", sex="XY", bam=b_tumor)
normal <- CamSample(id="N", sex="XY", bam=b_normal)
config <- CamConfig(outdir="./results", bsseq="wgbs", lib="pe", build="hg38", n_cores=10)
```

When defining `CamSample` objects, samples must have a unique `id` value between runs otherwise data will be overwritten. Alternatively, users may use the `patient` argument to separate samples the same `id`.

To run the pipeline:

```
pipeline(
    tumor=tumor,
    germline=normal,
    infiltrates=normal,
    origin=normal,
    config=config
)
```

### Use external CNA solutions

The germline sample is optional as, in the absence of patient-matched methylation data, you may already have an allele-specific CNA solutions for your bulk tumor. For example, this could be derived from bulk WGS of the same sampl.

You can provide this data in tab-delimited text file as shown below. Importantly,:

- column names are optional
- purity and ploidy values are taken from the first data row alone
- chromosome names may be given with or without 'chr' prefix

| chrom | start | end | major_cn | minor_cn | purity | ploidy |
| --------- | ---- | ---- | ---- | ---- | ---- | ---- |
| chr1 | 1 | 400 | 2 | 1 | 0.67 | 3.5 |
| chr1 | 401 | 1000 | 1 | 1 | 0.67 | 3.5 |

To run CAMDAC with this CNA solution, pass attach the file to the tumor `CamSample()` object:

```{r add_cna}
library(CAMDAC)

# Load test data
b_tumor <- system.file("testdata", "tumor.bam", package = "CAMDAC")
b_normal <- system.file("testdata", "normal.bam", package = "CAMDAC")
cna_file <- system.file("testdata", "test.cna.txt", package = "CAMDAC")

# Set config
config <- CamConfig(outdir="./results", bsseq="wgbs", lib="pe", build="hg38", n_cores=10)

# Create tumor object and attach CNA solution
tumor <- CamSample(id="T", sex="XY", bam=b_tumor)
attach_output(tumor, config, "cna", cna_file)

# Define normal object(s) for deconvolution or differential methylation
normal <- CamSample(id="N", sex="XY", bam=b_normal)

# Run pipeline with CNA solution
pipeline(
    tumor=tumor,
    germline=NULL,
    infiltrates=normal,
    origin=normal,
    config=config
)
```

Next, see `vignette("output")` for a detailed summary of CAMDAC results files.
