---
title: "questions"
output: rmarkdown::html_vignette
vignette: >
  %\VignetteIndexEntry{questions}
  %\VignetteEngine{rmarkdown::render}
  %\VignetteEncoding{UTF-8}
---

```{r, include = FALSE}
knitr::opts_chunk$set(
  collapse = TRUE,
  comment = "#>",
  eval=FALSE
)
```

```{r setup}
library(CAMDAC)
```

## Setup

### Are there external dependencies?

CAMDAC runs Battenberg to call copy number variants in WGBS mode. This requires requires the `java` command-line utility to be available in the system path. Otherwise, you can download Java from [https://openjdk.org/](https://openjdk.org/).

### How does CAMDAC find pipeline files?

CAMDAC searches for pipeline files in the following order:

1. A directory passed when creating the config object (see `CamConfig()`)
1. The location defined by the environment variable CAMDAC_PIPELINE_FILES.
1. The current working directory

We recommend that you set the environment variable `CAMDAC_PIPELINE_FILES` to the directory where you downloaded the files. This will allow CAMDAC to find the files automatically whenever you load R.

From a Unix terminal:
> echo "CAMDAC_PIPELINE_FILES=$(realpath R)" >> ~/.Renviron

### I have an error when downloading pipeline files

Some instutites setup firewalls for direct donwloads. If the download command fails, please [raise an issue on github](https://github.com/VanLoo-lab/CAMDAC/issues) describing the error. You can manually download the files from the endpoints listed in `inst/extdata/pipeline_files_urls.txt`. The files are stored as tar.gz files, therefore you will need to unpack them with the command `tar -xvzf <file>`.

## General

### What if I don't have a CNA or matched germline sample?

Unfortunately, CAMDAC cannot be run without a CNA solution or a normal sample from which to derive one.

### I have unmatched normal samples from multiple patients. Can I use CAMDAC to merge them?

Yes, please follow see the vignette `vignettes("panel")`.

### I want to run CAMDAC on something other than hg19 or hg38

Please raise an issue on GitHub to request files for a new reference genome.

## Extra features

### Can I skip steps of the analysis?

When calling `pipeline` if you do not give a normal infiltrate or cell of origin, the pipeline skips deconvolution and differential methylation respectively. This may be useful to run a quick first-pass to find and refit copy number solutions. When CAMDAC has found a solution and is rerun with the same tumor, config, and normal, the `infiltrates` and `cell_of_origin` arguments will continue the pipeline where it left off. The entire pipeline can be re-run be deleting the output directory or setting `overwrite=TRUE` in the `CamConfig`.

### My CNA solution was imperfect. How can I refit with different purity and ploidy values?

If you want to use an external purity and ploidy solution, simply pass a CNA file that has only the purity and ploidy fields. Additionally, set `refit==TRUE` in the CamConfig and CAMDAC will use this to refit the sample.

## I am only interested in specific regions. Can I limit my analysis using a BED file?

To analyse specific genomic regions, you may pass a BED file to CAMDAC config:

```
CamConfig(outdir=".", ref="./pipeline_files", regions="regions.bed")
```

CAMDAC will merge any overlapping regions prior to analysis.
