<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en">
<head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1.0">
<title>allele-specific • CAMDAC</title>
<!-- jquery --><script src="https://cdnjs.cloudflare.com/ajax/libs/jquery/3.4.1/jquery.min.js" integrity="sha256-CSXorXvZcTkaix6Yvo6HppcZGetbYMGWSFlBw8HfCJo=" crossorigin="anonymous"></script><!-- Bootstrap --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/css/bootstrap.min.css" integrity="sha256-bZLfwXAP04zRMK2BjiO8iu9pf4FbLqX6zitd+tIvLhE=" crossorigin="anonymous">
<script src="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/js/bootstrap.min.js" integrity="sha256-nuL8/2cJ5NDSSwnKD8VqreErSWHtnEP9E7AySL+1ev4=" crossorigin="anonymous"></script><!-- bootstrap-toc --><link rel="stylesheet" href="../bootstrap-toc.css">
<script src="../bootstrap-toc.js"></script><!-- Font Awesome icons --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/all.min.css" integrity="sha256-mmgLkCYLUQbXn0B1SRqzHar6dCnv9oZFPEC1g1cwlkk=" crossorigin="anonymous">
<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/v4-shims.min.css" integrity="sha256-wZjR52fzng1pJHwx4aV2AO3yyTOXrcDW7jBpJtTwVxw=" crossorigin="anonymous">
<!-- clipboard.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.6/clipboard.min.js" integrity="sha256-inc5kl9MA1hkeYUt+EC3BhlIgyp/2jDIyBLS6k3UxPI=" crossorigin="anonymous"></script><!-- headroom.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/headroom.min.js" integrity="sha256-AsUX4SJE1+yuDu5+mAVzJbuYNPHj/WroHuZ8Ir/CkE0=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script><!-- pkgdown --><link href="../pkgdown.css" rel="stylesheet">
<script src="../pkgdown.js"></script><meta property="og:title" content="allele-specific">
<meta property="og:description" content="CAMDAC">
<!-- mathjax --><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]-->
</head>
<body data-spy="scroll" data-target="#toc">
    

    <div class="container template-article">
      <header><div class="navbar navbar-default navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar" aria-expanded="false">
        <span class="sr-only">Toggle navigation</span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <span class="navbar-brand">
        <a class="navbar-link" href="../index.html">CAMDAC</a>
        <span class="version label label-default" data-toggle="tooltip" data-placement="bottom" title="">0.1.1</span>
      </span>
    </div>

    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav">
<li>
  <a href="../reference/index.html">Reference</a>
</li>
<li class="dropdown">
  <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" aria-expanded="false">
    Articles
     
    <span class="caret"></span>
  </a>
  <ul class="dropdown-menu" role="menu">
<li>
      <a href="../articles/allele-specific.html">allele-specific</a>
    </li>
    <li>
      <a href="../articles/copy_number_alt.html">copy_number</a>
    </li>
    <li>
      <a href="../articles/output.html">output</a>
    </li>
    <li>
      <a href="../articles/panels.html">panels</a>
    </li>
    <li>
      <a href="../articles/pipeline.html">pipeline</a>
    </li>
    <li>
      <a href="../articles/questions.html">questions</a>
    </li>
    <li>
      <a href="../articles/setup.html">setup</a>
    </li>
    <li>
      <a href="../articles/visualisation.html">visualisation</a>
    </li>
  </ul>
</li>
<li>
  <a href="../news/index.html">Changelog</a>
</li>
      </ul>
<ul class="nav navbar-nav navbar-right"></ul>
</div>
<!--/.nav-collapse -->
  </div>
<!--/.container -->
</div>
<!--/.navbar -->

      

      </header><div class="row">
  <div class="col-md-9 contents">
    <div class="page-header toc-ignore">
      <h1 data-toc-skip>allele-specific</h1>
            
      
      
      <div class="hidden name"><code>allele-specific.Rmd</code></div>

    </div>

    
    
<div class="section level2">
<h2 id="allele-specific-methylation-asm-analysis"> Allele-specific methylation (ASM) analysis<a class="anchor" aria-label="anchor" href="#allele-specific-methylation-asm-analysis"></a>
</h2>
<p>CAMDAC can be used to detect allele-specific methylation (ASM) by phasing CpGs to heterozygous SNPs and deconvolving bulk methylation rates per allele.</p>
<p>This tutorial steps through the ASM analysis pipeline. Briefly: 1. Count CpG methylation on tumor and normal at sites phased to SNP loci. 2. Deconvolve methylation on tumor <strong>per haplotype</strong> using the normal 3. Assign allele-specific copy number state <strong>per CpG</strong> using the bulk tumor solution 4. Call allele-specific differential methylation within samples 5. Call allele-specific differential methylation between samples</p>
<p>Results from this pipeline are found in the results directory under ‘PATIENT/AlleleSpecific’. The following files are generated: - <em>asm_counts.csv.gz - The number of reads supporting each allele at each CpG - </em>asm_hap_stats.csv.gz - Summary statistics for each phased SNP - <em>asm_phase_map.csv.gz - A mapping if read-CpG-SNP phasing - </em>snps.txt - The heterozygous SNP loci input for ASM analysis - <em>cna.csv - The tumor allele-specific copy number solution. See format in <code>vignettes("pipeline")</code>. - </em>asm_cg_cn.csv - Phased CpG copy number state - <em>asm_m.csv - Phased CpG bulk methylatio - </em>asm_m_pure.csv - Phased CpG pure methylation - <em>asm_m_within_diff.csv - Within-sample differential methylation - </em>asm_m_between_diff.csv - Between-sample differential methylation - *asm_meta.csv - Description of all input parameters for CAMDAC ASM run.</p>
</div>
<div class="section level2">
<h2 id="camdac-asm-from-bam-files">CAMDAC-ASM from BAM files<a class="anchor" aria-label="anchor" href="#camdac-asm-from-bam-files"></a>
</h2>
<p>The <code><a href="../reference/asm_pipeline.html">asm_pipeline()</a></code> function runs CAMDAC-ASM analysis by generates the allele-specific copy number solution and heterozygous SNP loci, followesd by deconvolution and differential ASM analysis:</p>
<div class="sourceCode" id="cb1"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">b_tumor</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/system.file.html" class="external-link">system.file</a></span><span class="op">(</span><span class="st">"testdata"</span>, <span class="st">"tumor.bam"</span>, package <span class="op">=</span> <span class="st">"CAMDAC"</span><span class="op">)</span></span>
<span><span class="va">b_normal</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/system.file.html" class="external-link">system.file</a></span><span class="op">(</span><span class="st">"testdata"</span>, <span class="st">"normal.bam"</span>, package <span class="op">=</span> <span class="st">"CAMDAC"</span><span class="op">)</span></span>
<span></span>
<span><span class="va">tumor</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/CamSample.html">CamSample</a></span><span class="op">(</span>id<span class="op">=</span><span class="st">"T"</span>, sex<span class="op">=</span><span class="st">"XY"</span>, bam<span class="op">=</span><span class="va">b_tumor</span><span class="op">)</span></span>
<span><span class="va">normal</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/CamSample.html">CamSample</a></span><span class="op">(</span>id<span class="op">=</span><span class="st">"N"</span>, sex<span class="op">=</span><span class="st">"XY"</span>, bam<span class="op">=</span><span class="va">b_normal</span><span class="op">)</span></span>
<span><span class="va">config</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/CamConfig.html">CamConfig</a></span><span class="op">(</span>outdir<span class="op">=</span><span class="st">"./results"</span>, bsseq<span class="op">=</span><span class="st">"wgbs"</span>, lib<span class="op">=</span><span class="st">"pe"</span>, build<span class="op">=</span><span class="st">"hg38"</span>, n_cores<span class="op">=</span><span class="fl">10</span><span class="op">)</span></span>
<span></span>
<span><span class="fu"><a href="../reference/asm_pipeline.html">asm_pipeline</a></span><span class="op">(</span></span>
<span>    tumor<span class="op">=</span><span class="va">tumor</span>,</span>
<span>    germline<span class="op">=</span><span class="va">normal</span>,</span>
<span>    infiltrates<span class="op">=</span><span class="va">normal</span>,</span>
<span>    origin<span class="op">=</span><span class="va">normal</span>,</span>
<span>    config<span class="op">=</span><span class="va">config</span></span>
<span><span class="op">)</span></span></code></pre></div>
</div>
<div class="section level2">
<h2 id="the-asm-results-are-placed-in-a-patientallelespecific-directory-with-the-following-files">The ASM results are placed in a PATIENT/AlleleSpecific directory with the following files:<a class="anchor" aria-label="anchor" href="#the-asm-results-are-placed-in-a-patientallelespecific-directory-with-the-following-files"></a>
</h2>
</div>
<div class="section level2">
<h2 id="todo-camdac-asm-from-external-inputs">TODO: CAMDAC-ASM from external inputs<a class="anchor" aria-label="anchor" href="#todo-camdac-asm-from-external-inputs"></a>
</h2>
<p>To run the ASM pipeline without BAM files, CAMDAC requires: - Each CamSample object has SNP loci - The tumor CamSample object has an allele-specific CNA solution - All CamSample objects have BAM files available for phasing</p>
<p>CAMDAC-ASM requires a file of heterozygous SNP loci against which CpGs will be phased. This is a tab-delimited file with a header containing four fields:</p>
<table class="table">
<thead><tr class="header">
<th>Field</th>
<th>Description</th>
</tr></thead>
<tbody>
<tr class="odd">
<td>chrom</td>
<td>Chromosome name</td>
</tr>
<tr class="even">
<td>pos</td>
<td>SNP loci position</td>
</tr>
<tr class="odd">
<td>ref</td>
<td>The reference allele (A/C/T/G)</td>
</tr>
<tr class="even">
<td>alt</td>
<td>The alternate SNP allele (A/C/T/G)</td>
</tr>
</tbody>
</table>
</div>
<div class="section level2">
<h2 id="asm-analysis-after-camdac-main-pipeline">ASM analysis after CAMDAC main pipeline<a class="anchor" aria-label="anchor" href="#asm-analysis-after-camdac-main-pipeline"></a>
</h2>
<p>TODO</p>
</div>
<div class="section level2">
<h2 id="asm-analysis-with-external-inputs">ASM analysis with external inputs<a class="anchor" aria-label="anchor" href="#asm-analysis-with-external-inputs"></a>
</h2>
<p>First, attach your SNP loci file to the tumor object with <code><a href="../reference/attach_output.html">attach_output()</a></code>, then run <code><a href="../reference/asm_pipeline.html">asm_pipeline()</a></code>:</p>
<div class="sourceCode" id="cb2"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># Setup CAMDAC samples</span></span>
<span><span class="va">tumor</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/CamSample.html">CamSample</a></span><span class="op">(</span>id<span class="op">=</span><span class="st">"tumor"</span>, sex<span class="op">=</span><span class="st">"XY"</span>, bam<span class="op">=</span><span class="va">b_tumor</span><span class="op">)</span></span>
<span><span class="va">normal</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/CamSample.html">CamSample</a></span><span class="op">(</span>id<span class="op">=</span><span class="st">"normal"</span>, sex<span class="op">=</span><span class="st">"XY"</span>, bam<span class="op">=</span><span class="va">b_normal</span><span class="op">)</span></span>
<span><span class="va">config</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/CamConfig.html">CamConfig</a></span><span class="op">(</span>outdir<span class="op">=</span><span class="st">"./results"</span>, ref<span class="op">=</span><span class="st">"./pipeline_files"</span>, bsseq<span class="op">=</span><span class="st">"wgbs"</span>, lib<span class="op">=</span><span class="st">"pe"</span>, cores<span class="op">=</span><span class="fl">10</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># Add SNPs</span></span>
<span><span class="va">asm_snps_file</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/system.file.html" class="external-link">system.file</a></span><span class="op">(</span><span class="st">"testdata"</span>, <span class="st">"test_het_snps.tsv"</span>, package <span class="op">=</span> <span class="st">"CAMDAC"</span><span class="op">)</span></span>
<span><span class="fu"><a href="../reference/attach_output.html">attach_output</a></span><span class="op">(</span><span class="va">tumor</span>, <span class="va">config</span>, <span class="st">"asm_snps"</span>, <span class="va">asm_snps_file</span><span class="op">)</span></span>
<span><span class="fu"><a href="../reference/attach_output.html">attach_output</a></span><span class="op">(</span><span class="va">normal</span>, <span class="va">config</span>, <span class="st">"asm_snps"</span>, <span class="va">asm_snps_file</span><span class="op">)</span></span></code></pre></div>
<p>Next, CAMDAC requires the allele-specific copy number solution from the tumor, attached as follows:</p>
<div class="sourceCode" id="cb3"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">cna_file</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/system.file.html" class="external-link">system.file</a></span><span class="op">(</span><span class="st">"testdata"</span>, <span class="st">"test_cna.tsv"</span>, package <span class="op">=</span> <span class="st">"CAMDAC"</span><span class="op">)</span></span>
<span><span class="fu"><a href="../reference/attach_output.html">attach_output</a></span><span class="op">(</span><span class="va">tumor</span>, <span class="va">config</span>, <span class="st">"cna"</span>, <span class="va">cna_file</span><span class="op">)</span></span></code></pre></div>
<p>Finally, run the allele-specific methylation pipeline:</p>
<div class="sourceCode" id="cb4"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="../reference/asm_pipeline.html">asm_pipeline</a></span><span class="op">(</span></span>
<span>  tumor<span class="op">=</span><span class="va">tumor</span>,</span>
<span>  infiltrates<span class="op">=</span><span class="va">normal</span>,</span>
<span>  origin<span class="op">=</span><span class="va">normal</span>,</span>
<span>  config<span class="op">=</span><span class="va">config</span></span>
<span><span class="op">)</span></span></code></pre></div>
</div>
<div class="section level2">
<h2 id="using-snps-called-from-previous-camdac-runs">Using SNPs called from previous CAMDAC runs<a class="anchor" aria-label="anchor" href="#using-snps-called-from-previous-camdac-runs"></a>
</h2>
<p>If you have already run the CAMDAC pipeline in tumor-normal mode, then the tumor object’s SNP files will be used by default.</p>
</div>
  </div>

  <div class="col-md-3 hidden-xs hidden-sm" id="pkgdown-sidebar">

        <nav id="toc" data-toggle="toc"><h2 data-toc-skip>Contents</h2>
    </nav>
</div>

</div>



      <footer><div class="copyright">
  <p></p>
<p>Developed by Elizabeth Larose Cadieux, Nana Mensah, Carla Castignani, Jonas Demeulemeester, Peter Van Loo.</p>
</div>

<div class="pkgdown">
  <p></p>
<p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.0.6.9000.</p>
</div>

      </footer>
</div>

  


  

  </body>
</html>
