<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en">
<head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1.0">
<title>deconvolution â€¢ CAMDAC</title>
<!-- jquery --><script src="https://cdnjs.cloudflare.com/ajax/libs/jquery/3.4.1/jquery.min.js" integrity="sha256-CSXorXvZcTkaix6Yvo6HppcZGetbYMGWSFlBw8HfCJo=" crossorigin="anonymous"></script><!-- Bootstrap --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/css/bootstrap.min.css" integrity="sha256-bZLfwXAP04zRMK2BjiO8iu9pf4FbLqX6zitd+tIvLhE=" crossorigin="anonymous">
<script src="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/js/bootstrap.min.js" integrity="sha256-nuL8/2cJ5NDSSwnKD8VqreErSWHtnEP9E7AySL+1ev4=" crossorigin="anonymous"></script><!-- bootstrap-toc --><link rel="stylesheet" href="../bootstrap-toc.css">
<script src="../bootstrap-toc.js"></script><!-- Font Awesome icons --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/all.min.css" integrity="sha256-mmgLkCYLUQbXn0B1SRqzHar6dCnv9oZFPEC1g1cwlkk=" crossorigin="anonymous">
<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/v4-shims.min.css" integrity="sha256-wZjR52fzng1pJHwx4aV2AO3yyTOXrcDW7jBpJtTwVxw=" crossorigin="anonymous">
<!-- clipboard.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.6/clipboard.min.js" integrity="sha256-inc5kl9MA1hkeYUt+EC3BhlIgyp/2jDIyBLS6k3UxPI=" crossorigin="anonymous"></script><!-- headroom.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/headroom.min.js" integrity="sha256-AsUX4SJE1+yuDu5+mAVzJbuYNPHj/WroHuZ8Ir/CkE0=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script><!-- pkgdown --><link href="../pkgdown.css" rel="stylesheet">
<script src="../pkgdown.js"></script><meta property="og:title" content="deconvolution">
<meta property="og:description" content="CAMDAC">
<!-- mathjax --><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]-->
</head>
<body data-spy="scroll" data-target="#toc">
    

    <div class="container template-article">
      <header><div class="navbar navbar-default navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar" aria-expanded="false">
        <span class="sr-only">Toggle navigation</span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <span class="navbar-brand">
        <a class="navbar-link" href="../index.html">CAMDAC</a>
        <span class="version label label-default" data-toggle="tooltip" data-placement="bottom" title="">0.1.1</span>
      </span>
    </div>

    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav">
<li>
  <a href="../reference/index.html">Reference</a>
</li>
<li class="dropdown">
  <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" aria-expanded="false">
    Articles
     
    <span class="caret"></span>
  </a>
  <ul class="dropdown-menu" role="menu">
<li>
      <a href="../articles/deconvolution.html">deconvolution</a>
    </li>
  </ul>
</li>
<li>
  <a href="../news/index.html">Changelog</a>
</li>
      </ul>
<ul class="nav navbar-nav navbar-right"></ul>
</div>
<!--/.nav-collapse -->
  </div>
<!--/.container -->
</div>
<!--/.navbar -->

      

      </header><div class="row">
  <div class="col-md-9 contents">
    <div class="page-header toc-ignore">
      <h1 data-toc-skip>deconvolution</h1>
            
      
      
      <div class="hidden name"><code>deconvolution.Rmd</code></div>

    </div>

    
    
<div class="section level2">
<h2 id="introduction">Introduction<a class="anchor" aria-label="anchor" href="#introduction"></a>
</h2>
<p>Here, we demonstrate the simplest case for CAMDAC deconvolution using a bulk tumor with adjacent normal tissue from the same patient. We can define CAMDAC inputs using two functions:</p>
<ul>
<li>
<code><a href="../reference/create_camdac_config.html">create_camdac_config()</a></code> sets the configuration for camdac pipeline functions.</li>
<li>
<code><a href="../reference/create_camdac_sample.html">create_camdac_sample()</a></code> sets details for each biological sample.</li>
</ul>
<p>After which, we pass these input objects to <code>pipline_*</code> functions. This vignette will walk you through running the <code><a href="../reference/pipeline_tumor_normal.html">pipeline_tumor_normal()</a></code> workflow.</p>
<p>Load the CAMDAC library:</p>
<div class="sourceCode" id="cb1"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/base/message.html" class="external-link">suppressPackageStartupMessages</a></span><span class="op">(</span></span>
<span>  <span class="fu"><a href="https://rdrr.io/r/base/warning.html" class="external-link">suppressWarnings</a></span><span class="op">(</span></span>
<span>    <span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va">CAMDAC</span><span class="op">)</span></span>
<span>  <span class="op">)</span></span>
<span><span class="op">)</span></span></code></pre></div>
</div>
<div class="section level2">
<h2 id="configuration">Configuration<a class="anchor" aria-label="anchor" href="#configuration"></a>
</h2>
<div class="section level3">
<h3 id="reference-files">Reference files<a class="anchor" aria-label="anchor" href="#reference-files"></a>
</h3>
<p>CAMDAC requires a set of pre-built reference files. These files are stored at a dropbox endpoint as tar.gz files. You can download these files with the command <code>CAMDAC::download_pipeline_files("wgbs")"</code>.</p>
<p>CAMDAC pipeline functions search for the unpacked reference files in the following locations:</p>
<ol style="list-style-type: decimal">
<li>The directory passed when creating the CAMDAC config object (see camdac_refs argument in <code>create_camdac_config</code>)</li>
<li>The location defined by the environment variable CAMDAC_PIPELINE_FILES.</li>
<li>The current working directory</li>
</ol>
<p>If the download command fails, you can manually download the reference files from the endpoints listed in <code>inst/extdata/pipeline_files_urls.txt</code>.</p>
</div>
<div class="section level3">
<h3 id="configuration-object">Configuration object<a class="anchor" aria-label="anchor" href="#configuration-object"></a>
</h3>
<p>Once reference files are downloaded, load the CAMDAC library and set up the configuration:</p>
<div class="sourceCode" id="cb2"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">config</span> <span class="op">=</span> <span class="fu"><a href="../reference/create_camdac_config.html">create_camdac_config</a></span><span class="op">(</span></span>
<span>  <span class="co"># camdac_refs = "./camdac_refs", # Optional: Pass the location of pipeline files </span></span>
<span>  outdir<span class="op">=</span><span class="st">"./results"</span>, <span class="co"># Path for outputs</span></span>
<span>  bsseq<span class="op">=</span><span class="st">"wgbs"</span>, <span class="co"># WGBS</span></span>
<span>  build<span class="op">=</span><span class="st">"hg38"</span>, <span class="co"># Reference</span></span>
<span>  bsseq_lib<span class="op">=</span><span class="st">"pe"</span>, <span class="co"># Paired end</span></span>
<span>  n_cores <span class="op">=</span> <span class="fl">10</span> <span class="co"># Cores for parallel processing</span></span>
<span><span class="op">)</span></span></code></pre></div>
</div>
</div>
<div class="section level2">
<h2 id="deconvolution">Deconvolution<a class="anchor" aria-label="anchor" href="#deconvolution"></a>
</h2>
<p>Next, define CAMDAC samples from the bulk tumor and normal BAM files. Test BAM files are provided with CAMDAC to verify your installation but please be aware that <strong>these are not whole genome bisulfite sequencing data</strong>, therefore the outputs from these files are not meaningful.</p>
<div class="sourceCode" id="cb3"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># Get test BAM</span></span>
<span><span class="va">tumor_bam</span> <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/system.file.html" class="external-link">system.file</a></span><span class="op">(</span><span class="st">"extdata"</span>, <span class="st">"NA18939_test_v3.bam"</span>, package<span class="op">=</span><span class="st">"CAMDAC"</span><span class="op">)</span></span>
<span><span class="va">normal_bam</span> <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/system.file.html" class="external-link">system.file</a></span><span class="op">(</span><span class="st">"extdata"</span>, <span class="st">"NA20502_test_v3.bam"</span>, package<span class="op">=</span><span class="st">"CAMDAC"</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># Set CAMDAC tumor sample</span></span>
<span><span class="va">tumour</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/create_camdac_sample.html">create_camdac_sample</a></span><span class="op">(</span></span>
<span>  patient_id<span class="op">=</span><span class="st">"P1"</span>,</span>
<span>  patient_sex<span class="op">=</span><span class="st">"XX"</span>,</span>
<span>  sample_id<span class="op">=</span><span class="st">"T"</span>,</span>
<span>  sample_type<span class="op">=</span><span class="st">"tumour"</span>,</span>
<span>  bam_file<span class="op">=</span><span class="va">tumor_bam</span></span>
<span><span class="op">)</span></span>
<span></span>
<span><span class="co"># Set CAMDAC normal sample</span></span>
<span><span class="va">normal</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/create_camdac_sample.html">create_camdac_sample</a></span><span class="op">(</span></span>
<span>  patient_id<span class="op">=</span><span class="st">"P1"</span>,</span>
<span>  patient_sex<span class="op">=</span><span class="st">"XX"</span>,</span>
<span>  sample_id<span class="op">=</span><span class="st">"N"</span>,</span>
<span>  sample_type<span class="op">=</span><span class="st">"normal"</span>,</span>
<span>  bam_file<span class="op">=</span><span class="va">normal_bam</span></span>
<span><span class="op">)</span></span></code></pre></div>
<p>Run the CAMDAC tumor-normal pipeline. This analysis takes 1 hour using the test dataset, and approximately 8 hours on a full WGBS dataset.</p>
<div class="sourceCode" id="cb4"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="../reference/pipeline_tumor_normal.html">pipeline_tumor_normal</a></span><span class="op">(</span><span class="va">tumor</span>, <span class="va">normal</span>, <span class="va">config</span><span class="op">)</span></span></code></pre></div>
<p>In this pipeline, the adjacent normal methylation profile is used to select germline heterozygous SNPs for copy number calling, to estimate the infiltrating cell methylation, and as a proxy for the cell of origin in differential methylation calling.</p>
</div>
<div class="section level2">
<h2 id="results">Results<a class="anchor" aria-label="anchor" href="#results"></a>
</h2>
<p>CAMDAC outputs are written in the directory given by <code>config$outdir</code> in the format <code>PATIENT/DATASET/SAMPLE/</code>:</p>
<pre><code>â””â”€â”€ P1
    â”œâ”€â”€ Allelecounts
    â”‚   â”œâ”€â”€ N
    â”‚   â”‚   â””â”€â”€ P1.N.SNPs.CpGs.all.sorted.csv.gz
    â”‚   â””â”€â”€ T
    â”‚       â””â”€â”€ P1.T.SNPs.CpGs.all.sorted.csv.gz
    â”œâ”€â”€ Copynumber
    â”‚   â””â”€â”€ T
    â”‚       â”œâ”€â”€ battenberg # (Subset of files shown)
    â”‚           â”œâ”€â”€ P1-T_subclones.txt
    â”‚           â””â”€â”€ P1-T_BattenbergProfile_average.png
    â”‚       â”œâ”€â”€ P1.T.ACF.and.ploidy.txt
    â”‚       â”œâ”€â”€ P1.T.ascat.bc.qs
    â”‚       â”œâ”€â”€ P1.T.ascat.frag.qs
    â”‚       â”œâ”€â”€ P1.T.ascat.output.qs
    â”‚       â”œâ”€â”€ P1.T.ASPCF.png
    â”‚       â”œâ”€â”€ P1.T.germline.png
    â”‚       â”œâ”€â”€ P1.T.SNPs.csv.gz
    â”‚       â”œâ”€â”€ P1.T.ASCATprofile.png
    â”‚       â”œâ”€â”€ P1.T.BAF.PCFed.txt
    â”‚       â”œâ”€â”€ P1.T.LogR.PCFed.txt
    â”‚       â”œâ”€â”€ P1.T.rawprofile.png
    â”‚       â”œâ”€â”€ P1.T.sunrise.png
    â”‚       â””â”€â”€ P1.T.tumour.png
    â””â”€â”€ Methylation
        â”œâ”€â”€ N
        â”‚   â””â”€â”€ P1.N.m.csv.gz
        â””â”€â”€ T
            â”œâ”€â”€ P1.T.CAMDAC_annotated_DMRs.fst
            â”œâ”€â”€ P1.T.CAMDAC_results_per_CpG.fst
            â”œâ”€â”€ P1.T.m.csv.gz
            â””â”€â”€ P1.T.pure.qs
</code></pre>
</div>
  </div>

  <div class="col-md-3 hidden-xs hidden-sm" id="pkgdown-sidebar">

        <nav id="toc" data-toggle="toc"><h2 data-toc-skip>Contents</h2>
    </nav>
</div>

</div>



      <footer><div class="copyright">
  <p></p>
<p>Developed by Elizabeth Larose Cadieux, Nana Mensah, Carla Castignani, Jonas Demeulemeester, Peter Van Loo.</p>
</div>

<div class="pkgdown">
  <p></p>
<p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.0.6.9000.</p>
</div>

      </footer>
</div>

  


  

  </body>
</html>
